external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java

```diff
@@ -0,0 +1,449 @@
+                    BigInteger ackedSequenceNumber = ackedSequenceNumbers.next();
+                    if (commitSequenceNumberBound == null || (commitSequenceNumberBound.compareTo(ackedSequenceNumber) == 1)) {
+                        ackedSequenceNumberToCommit = ackedSequenceNumber;
+                        ackedSequenceNumbers.remove();
+                    } else {
+                        break;
+                    }
+                }
+                if (ackedSequenceNumberToCommit != null) {
+                    Map<Object, Object> state = new HashMap<>();
+                    state.put("committedSequenceNumber", ackedSequenceNumberToCommit.toString());
+                    LOG.debug("Committing sequence number " + ackedSequenceNumberToCommit.toString() + " for shardId " + shardId);
+                    zkConnection.commitState(kinesisConfig.getStreamName(), shardId, state);
+                }
+            }
+        }
+        lastCommitTime = System.currentTimeMillis();
+    }
+
+    void activate () {
+        LOG.info("Activate called");
+        deactivated = false;
+        kinesisConnection.initialize();
+    }
+
+    void deactivate () {
+        LOG.info("Deactivate called");
+        deactivated = true;
+        commit();
+        kinesisConnection.shutdown();
+    }
+
+    void close () {
+        commit();
+        kinesisConnection.shutdown();
+        zkConnection.shutdown();
+    }
+
+    // fetch records from kinesis starting at sequence number for message passed as argument. Any other messages fetched and are in the failed queue will also
+    // be kept in memory to avoid going to kinesis again for retry
+    private void fetchFailedRecords (KinesisMessageId kinesisMessageId) {
+        // if shard iterator not present for this message, get it
+        if (!shardIteratorPerFailedMessage.containsKey(kinesisMessageId)) {
+            refreshShardIteratorForFailedRecord(kinesisMessageId);
+        }
+        String shardIterator = shardIteratorPerFailedMessage.get(kinesisMessageId);
+        LOG.debug("Fetching failed records for shard id :" + kinesisMessageId.getShardId() + " at sequence number " + kinesisMessageId.getSequenceNumber() +
+                " using shardIterator " + shardIterator);
+        try {
+            GetRecordsResult getRecordsResult = kinesisConnection.fetchRecords(shardIterator);
+            if (getRecordsResult != null) {
+                List<Record> records = getRecordsResult.getRecords();
+                LOG.debug("Records size from fetchFailedRecords is " + records.size());
+                // update the shard iterator to next one in case this fetch does not give the message.
+                shardIteratorPerFailedMessage.put(kinesisMessageId, getRecordsResult.getNextShardIterator());
+                if (records.size() == 0) {
+                    LOG.warn("No records returned from kinesis. Hence sleeping for 1 second");
+                    Thread.sleep(1000);
```

Comment:
Sorry missed these earlier. Any reason for it to sleep exactly one sec or should it be configurable.